### PR TITLE
PARQUET-1770: [C++][CI] Add fuzz target for reading Parquet files

### PR DESCRIPTION
**Pull request by pitrou**
This fuzz target goes through the Parquet Arrow file reader.

### CODE REVIEW

**Comment by pitrou** [comment]
cc @wesm for the Parquet additions.

**Comment by pitrou** [comment]
We will need to merge this PR before https://github.com/google/oss-fuzz/pull/3374 is merged.
(or, if this can't be done quickly, move the relevant fix to a separate JIRA and PR)

File: cpp/src/parquet/arrow/generate_fuzz_corpus.cc [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+static constexpr int32_t kBatchSize = 1000;
+static constexpr int32_t kChunkSize = kBatchSize * 3 / 8;
+
+Result<std::shared_ptr<RecordBatch>> ExampleBatch1() {
+  random::RandomArrayGenerator gen(42);
+  std::shared_ptr<Array> a, b, c, d;
+  a = gen.Int16(kBatchSize, -10000, 10000, /*null_probability=*/0.2);
+  b = gen.Float64(kBatchSize, -1e10, 1e10, /*null_probability=*/0.0);
+  c = gen.String(kBatchSize, 0, 30, /*null_probability=*/0.2);
```
Comment (fsaintjacques) [inline cpp/src/parquet/arrow/generate_fuzz_corpus.cc]: Beware that DICT_ENCODING in the writer logic is a threshold on the page size. I'd double check that we trigger the dict encoding path.

File: cpp/src/parquet/arrow/generate_fuzz_corpus.cc [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  {
+    auto values = gen.Int64(kBatchSize * 10, -10000, 10000, /*null_probability=*/0.2);
+    auto offsets = gen.Offsets(kBatchSize + 1, 0, static_cast<int32_t>(values->length()));
+    RETURN_NOT_OK(ListArray::FromArrays(*offsets, *values, default_memory_pool(), &d));
+  }
```
Comment (fsaintjacques) [inline cpp/src/parquet/arrow/generate_fuzz_corpus.cc]: Add an integer column with a constant value to trigger RLE.

File: cpp/src/parquet/arrow/reader_internal.cc [hunk 3]
```diff
@@ -873,7 +875,7 @@ Status TransferDate64(RecordReader* reader, MemoryPool* pool,
 Status TransferDictionary(RecordReader* reader,
                           const std::shared_ptr<DataType>& logical_value_type,
                           std::shared_ptr<ChunkedArray>* out) {
-  auto dict_reader = dynamic_cast<internal::DictionaryRecordReader*>(reader);
+  auto dict_reader = dynamic_cast<DictionaryRecordReader*>(reader);
```
Comment (fsaintjacques) [inline cpp/src/parquet/arrow/reader_internal.cc]: Might as well replace it with checked_cast.

**Review by fsaintjacques — APPROVED** [review]
LGTM, can't wait to see what it'll find :)

File: cpp/src/parquet/arrow/generate_fuzz_corpus.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/generate_fuzz_corpus.cc]: I wasn't even thinking about automatic dict encoding :-)

File: cpp/src/parquet/arrow/generate_fuzz_corpus.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/generate_fuzz_corpus.cc]: Is that automatic too?

File: cpp/src/parquet/arrow/reader_internal.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/reader_internal.cc]: Indeed.

File: cpp/src/parquet/arrow/generate_fuzz_corpus.cc (hunk 2 above)
Comment (fsaintjacques) [reply cpp/src/parquet/arrow/generate_fuzz_corpus.cc]: I think so, I haven't checked how to trigger it. A repeated constant should, hopefully!

**Comment by pitrou** [comment]
@fsaintjacques Can you take a quick look again?

File: cpp/src/parquet/arrow/reader_internal.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/reader_internal.cc]: Actually, no, the base class is virtual, so `dynamic_cast` is required.

**Comment by pitrou** [comment]
It seems all columns end up RLE-encoded (and most dictionary-encoded):
https://gist.github.com/pitrou/ea74a0efa495ff6cc9c948dc7e1caa93

**Comment by wesm** [comment]
@pitrou that's right -- you can disable dictionary encoding on a per-column basis, though

**Comment by wesm** [comment]
+1 (sorry for the late review)

**Pull request closed without merging**